### DISCUSSION TITLE
Singleton type services

### DISCUSSION THREAD

**Discussion by Arcalise08** [vercel/next.js#51897] [Help] [answered]
### Summary

I'm having trouble understanding exactly how to create services that the backend of nextjs can reuse. For example, Signalr connections or CosmosDB clients. I've tried using static properties to no avail. I've got it sort of working with use of the global keyword but i'm told that may not be good to use and i have no idea how to properly declare a type for the global keyword so i have to use @ts-ignore on everything. Some of the people i've talked to have said that every request in nextjs is executed and isolation so theres no way to do this?

Surely that can't be right as that would require a db client for every single user of the website which seems crazy.

**Comment by icyJoseph** [comment, marked as answer]
Well it depends... if your app is to be hosted in a serverless environment, such as Vercel, then yeah, it'll spin up a task, which spins up whatever its JavaScript says, and it is then killed. 

Otherwise, you can just use something like:

```js
let client;

const createClient = () => {
  if (client) return client;
  client = actuallyCreateClient();
  return client;
}

export default createClient;
```

It varies from tech stack to stack stack of course, but that's roughly the sketch.

**Reply by Arcalise08** [reply]
This is the correct answer just only works in production. Tested it out and it works. 

Thank you!

**Comment by Arcalise08** [comment]
In development it'd be ran with docker-compose with a few other microservices using dapr to communicate. I got a warning that nextjs was running out of memory which i think is because its spinning up dapr clients everytime a request is sent out which stacks up.

**Comment by Arcalise08** [comment]
I'd also mention doing what you said like so

```
let client : DaprClient | null;
const daprHost = process.env.daprHost!
const daprPort = process.env.daprPort!
const createDaprClient = () => {
    if (client)
        return client;
    console.log("spinning up dapr client");
    client = new DaprClient({daprHost: daprHost, daprPort: daprPort});
    return client;
}
```

![Screenshot 2023-06-27 133403](https://github.com/vercel/next.js/assets/61042848/9a2548eb-e20e-4aa9-a226-1b5c26b6a061)

I can see it loses client almost immediately.

**Reply by icyJoseph** [reply]
Right, but that's dev mode, there code gets re-compiled and what not. 

Try with a build, make sure pages stay dynamic, perhaps drop a `export const dynamic = 'force dynamic'` and log every time a client is created.
